Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp

```diff
@@ -128,8 +129,28 @@ Ref<PAL::WebGPU::ExternalTexture> RemoteDeviceProxy::importExternalTexture(const
     }
 
     auto identifier = WebGPUIdentifier::generate();
+#if PLATFORM(COCOA) && ENABLE(VIDEO)
+    if (auto pixelBuffer = descriptor.pixelBuffer) {
+        WebKit::SharedVideoFrameWriter sharedVideoFrameWriter;
+        constexpr bool canSendIOSurface = false;
+        auto sharedVideoFrameBuffer = sharedVideoFrameWriter.writeBuffer(pixelBuffer.get(), [&](auto& semaphore) {
+            auto sendResult = send(Messages::RemoteDevice::SetSharedVideoFrameSemaphore { semaphore });
+            UNUSED_VARIABLE(sendResult);
+        }, [&](auto&& handle) {
+            auto sendResult = send(Messages::RemoteDevice::SetSharedVideoFrameMemory { WTFMove(handle) });
+            UNUSED_VARIABLE(sendResult);
+        }, canSendIOSurface);
+
+        auto sendResult = send(Messages::RemoteDevice::ImportExternalTextureFromPixelBuffer(*convertedDescriptor, sharedVideoFrameBuffer, identifier));
+        UNUSED_VARIABLE(sendResult);
+    }  else {
+        auto sendResult = send(Messages::RemoteDevice::ImportExternalTexture(*convertedDescriptor, identifier));
```

Comment:
This is a strange pattern, SharedVideoFrame is supposed to handle all video frames, remote or not.
Why are we separating them? Usually we would send a single SharedVideoFrame object.

Also, creating a SharedVideoFrame for each video frame will mean that we will create a shared memory for each video frame, which is inefficient in case we do repetitive GPU processing of video frames of the same type/same size (say a MediaStreamTrack).

Also, I do not see any test for any of either the remote or local code path.

Reply:
I think I can use `SharedVideoFrame` as long as I pass the VideoFrame as an opaque pointer through the WebGPU PAL layer, which can not see WebCore types. I will try that instead.

Reply:
Regarding testing, we have some tests from the WebGPU CTS checked in at `./LayoutTests/http/tests/webgpu/cts.https.html`:
```
<meta name=variant content='webgpu:web_platform,external_texture,video:importExternalTexture,sample:*'>
<meta name=variant content='webgpu:web_platform,external_texture,video:importExternalTexture,expired:*'>
<meta name=variant content='webgpu:web_platform,external_texture,video:importExternalTexture,compute:*'>
```

these tests are not yet passing with these changes. I'm investigating why as part of https://bugs.webkit.org/show_bug.cgi?id=257724